sematic/log_reader.py

```diff
@@ -422,40 +505,78 @@ def line_stream_from_log_directory(
         key=lambda path_key: int(
             path_key.replace(directory, "").replace(".log", "".replace("/", ""))
         ),
+        reverse=reverse,
     )
-    found_cursor_file = cursor_file is None
-    found_cursor_line = cursor_line_index is None
-    for log_file in log_files:
-        if log_file == cursor_file:
-            found_cursor_file = True
-        if not found_cursor_file:
-            continue
+    if len(log_files) == 0:
+        yield from []
+        return
+    index_of_log_file = 0
+    if cursor_file is not None:
+        if cursor_file not in log_files:
+            raise ValueError(f"Cursor referenced invalid log file {cursor_file}.")
+        index_of_log_file = log_files.index(cursor_file)
+
+    cursor_file = log_files[index_of_log_file]
+    log_files_to_traverse = log_files[index_of_log_file:]
+    for log_file in log_files_to_traverse:
+        start_line_index = DEFAULT_END_INDEX if reverse else DEFAULT_START_INDEX
+        if log_file == cursor_file and cursor_line_index is not None:
+            start_line_index = cursor_line_index
+
         text_stream: Iterable[str] = storage.get_line_stream(log_file)
+        yield from _stream_from_text_stream_from_index(
+            log_file, start_line_index, reverse, text_stream
+        )
+
+
+def _stream_from_text_stream_from_index(
+    log_file: str, start_line_index: int, reverse: bool, text_stream: Iterable[str]
+) -> Iterable[LogLine]:
+    """Starting at the given index within the file, stream logs in desired direction.
+
+    When going forward, the line pointed to by the index will be included. When going in
+    the reverse direction, the line pointed to by the index will be omitted.
+    """
+    # For purposes of this function, "preceding" refers to indices
+    # lower than the start index, regardless of traversal direction.
+    lines_preceding_cursor: List[LogLine] = []
+    if reverse:
         for i_line, line in enumerate(text_stream):
-            if (
-                (not found_cursor_line)
-                and cursor_line_index is not None
-                and i_line < cursor_line_index
-            ):
-                continue
-            found_cursor_line = True
-            yield LogLine(
+            as_log_line = LogLine(
+                source_file=log_file,
+                source_file_index=i_line,
+                line=line,
+            )
+            if i_line < start_line_index:
```

Comment:
1. But we want the *last* 50 lines preceding the cursor, and not the *first* 50. So even if we have passed 50 lines that would pass the filter, we need to keep going until we reach the cursor. Then we can go backwards from there. The trouble is we aren't able to random-seek to the cursor, as the blob storage only allows seeking from the 0 byte, in the forwards direction. Luckily, all log files contain at most 1000 lines, so we never need to seek within a file too far.
2. Sure! One thing to be aware of is that most of the logic in this module uses python [generator functions](https://wiki.python.org/moin/Generators). Therefore in many cases though the code "looks" like it is doing something in bulk, it is actually streaming them. This function (`_stream_from_text_stream_from_index`) is streaming log lines from a line stream that originates in the blob storage, and is not yet filtered. The code that is pulling from the stream created by `_stream_from_text_stream_from_index` is [get_log_lines_from_line_stream](https://github.com/sematic-ai/sematic/blob/6f37027766097aeeec1798767466a78762826ced/sematic/log_reader.py#L600)

Reply:
1. Yeah. I also refer to the last 50 up to the cursor, not the first 50. Having to go through the ones before the 50th to the last of the cursor seemed like it could have been more productive. 

2 .I see. This looks great --- everything is in its best position. Thanks for sharing the code pointer. 

It feels a little pitiful that we dug the whole log file, did an O(n) traversal a couple of times, but potentially only took a small portion of it. Does it make sense to cache the S3 log file in memory? Maybe with an LRU cache?

If it doesn't make sense, is one reason that we always pull 2000 lines of logs every time?